### PR TITLE
add recursive rule

### PR DESCRIPTION
**Pull request by mpskex**
We add support for recursive rules.

For example:
```python
import re
from lexer import Lexer, Rule

text = 'a b(ac) c'
lexer = Lexer(rules=[Rule('a', re.compile('a\s*')),
                     Rule('b', re.compile('b\(\w+\)\s*'), next=[
                         Rule('clause', re.compile('b\(')),
                         Rule('obj', re.compile('\w+')),
                         Rule(')', re.compile('\)')),
                         ]),
                     Rule('c', re.compile('c')),
                     ])
t = lexer.lex(text)
```

This code will give you:
```
[
    Token('a', 'a', 0, 0),
    Token(
        'b',
        [
            Token('clause', 'b(', 0, 0),
            Token('obj', 'ac', 2, 0),
            Token(')', ')', 4, 0)
        ],
        5, 0
    ),
    Token('c', 'c', 7, 0)
]
```

### CODE REVIEW

**Comment by mristin** [comment]
@mpskex please apologize for the delay; it's been busy lately on my side. I'll try to review the pull request till this weekend.

**Comment by mristin** [comment]
@mpskex indeed, Python 3.6 is not supported by GitHub anymore. I'll try to fix the CI tonight.

File: lexery/__init__.py [hunk 1]
```diff
@@ -52,6 +52,26 @@ def __init__(self, identifier: str, pattern: Pattern) -> None:
...
+    def match(self, text, pos, lno):
+        mtch = self.pattern.match(text, pos)
+        ret = []
+        if self.next is not None and mtch is not None:
+            pos = 0
+            for r in self.next:
+                _mtch, _t = r.match(mtch.group(), pos, 0)
+                if _mtch:
+                    ret.append(_t)
+                    pos += len(_mtch.group())
```
Comment (mristin) [inline lexery/__init__.py:66]:
```suggestion
                another_mtch, another_t = r.match(mtch.group(), pos, 0)
                if another_mtch:
                    ret.append(another_t)
                    pos += len(another_mtch.group())
```

File: lexery/__init__.py [hunk 2]
```diff
@@ -190,11 +210,10 @@ def lex(self, text: str) -> List[List[Token]]:
             while position < len(line):
                 mtched = False
                 for rule in self.rules:
-                    mtch = rule.pattern.match(line, position)
+                    _line = line
```
Comment (mristin) [inline lexery/__init__.py:213]: Please rename `_line` to `another_line`. Ditto for `_token` to `another_token`.

We use `_*` naming convention for private/protected members of a class.

**Review by mristin — CHANGES_REQUESTED** [review]
Thanks a lot, @mpskex ! I made only a couple of minor change suggestions related to the style of the code.

**Comment by mristin** [comment]
@mpskex thanks again! I fixed CI in the meanwhile. Could you please rebase your fork on the new master and have a look at my review comments?

**Review by mristin — APPROVED** [review]
LGTM, thanks!

**Comment by mpskex** [comment]
I just reran `precommit.py` and found `pydocstyle` command was conflicting the format options. So I disabled `D203`, `D204` and `D212` in `pydocstyle` check. Now it should be good.

**Comment by mristin** [comment]
@mpskex the format still seems to be off? Could you please run `precommit.py --overwrite`, and commit?

**Comment by mpskex** [comment]
> I just reran `precommit.py` and found `pydocstyle` command was conflicting the format options. So I disabled `D203`, `D204` and `D212` in `pydocstyle` check. Now it should be good.

This is reverted.

**Comment by mpskex** [comment]
> @mpskex the format still seems to be off? Could you please run `precommit.py --overwrite`, and commit?

Sry for that, it must be the wrong yapf version i installed. The format should be ok now.

**Comment by mristin** [comment]
@mpskex thanks a lot! I'm just about to release a new minor version.

Please apologize for the delay. I was out-of-town with my family on holidays, and then had to deal with a backlog after coming back.

**Comment by mristin** [comment]
P.S. @mpskex if you haven't done so, please consider starring the project to give it some visibility.

**Pull request merged**